src/cargo/util/graph.rs

```diff
@@ -154,17 +324,251 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    pub fn add_edge(&mut self, node: N, child: N, edge: E) {
+        self.age = {
+            let mut g = self.activate();
+            g.add_edge(node, child, edge);
+            g.len()
+        };
+    }
+
+    pub fn contains<Q: ?Sized>(&self, k: &Q) -> bool
+    where
+        N: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.borrow()
+            .inner
+            .nodes
+            .get_full(k)
+            .filter(|(i, _, _)| *i < self.age.len_nodes)
+            .is_some()
+    }
+
+    /// Checks if there is a path from `from` to `to`.
+    pub fn is_path_from_to<'a>(&'a self, from: &'a N, to: &'a N) -> bool {
+        let mut stack = vec![from];
+        let mut seen = HashSet::new();
+        let inner = &self.borrow().inner;
+        seen.insert(from);
+        while let Some((_, _, iter)) = stack.pop().and_then(|p| {
+            inner
+                .nodes
+                .get_full(p)
+                .filter(|(i, _, _)| *i < self.age.len_nodes)
+        }) {
+            for (p, idx) in iter.iter() {
+                if *idx < self.age.len_edges {
+                    if p == to {
+                        return true;
+                    }
+                    if seen.insert(p) {
+                        stack.push(p);
+                    }
+                }
+            }
+        }
+        false
+    }
+}
+
+impl<N: Eq + Hash + Clone, E: Clone + PartialEq> Default for StackGraph<N, E> {
+    fn default() -> StackGraph<N, E> {
+        StackGraph::new()
+    }
+}
+
+impl<N: fmt::Display + Eq + Hash + Clone, E: Clone + fmt::Debug + PartialEq> fmt::Debug
+    for StackGraph<N, E>
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(fmt, "Graph {{")?;
+
+        for (from, e) in self.borrow().inner.nodes.iter().take(self.age.len_nodes) {
+            writeln!(fmt, "  - {}", from)?;
+
+            for (to, idx) in e.iter() {
+                if *idx < self.age.len_edges {
+                    writeln!(fmt, "    - {}", to)?;
+                    for edge in self.borrow().edge(from, to) {
+                        writeln!(fmt, "      - {:?}", edge)?;
+                    }
+                }
+            }
+        }
+
+        write!(fmt, "}}")?;
+
+        Ok(())
+    }
+}
+
+pub struct StackGraphView<'a, N: Clone, E: Clone> {
+    inner: Ref<'a, Graph<N, E>>,
+    age: GraphAge,
+}
+
+impl<'a, N: Eq + Hash + Clone, E: Clone + PartialEq> StackGraphView<'a, N, E> {
+    pub fn iter(&self) -> impl Iterator<Item = &N> {
+        self.inner.nodes.keys().take(self.age.len_nodes)
+    }
+
+    pub fn edge(&self, from: &N, to: &N) -> Edges<'_, E> {
+        Edges {
+            graph: &self.inner.edges[..self.age.len_edges],
+            index: self
+                .inner
+                .nodes
+                .get(from)
+                .and_then(|x| x.get(to).copied())
+                .filter(|idx| idx < &self.age.len_edges),
```

Comment:
The extra filtering done here in the view is somewhat odd, can you be sure to add comments (not sure where, just somewhere would be fine) as to why it's necessary?